### PR TITLE
Generators for iterating through file collection

### LINKED ISSUE #25: Generators for iterating through filecollection
Here is a rough start to defining special iterators for going through the dataframes from `filecollection`

## Goal

After using `filecollection` something I might want to do is create a master bias frame. I can do this easily by filtering the collection for the biases (OBJECT="")
```julia
full_coll = filecollection(...)
bias_coll = filter(r -> r.OBJECT == "", full_coll)
```
now, I want to be able to use the methods like `combine` using this dataframe! Here is a simple example-

```julia
bias_arr = map(images(bias_coll)) do img
    # img isa ImageHDU
	tmp = subtract_overscan(img, :BIASSEC)
	trim!(tmp, :BIASSEC)
end
master_bias = combine(bias_arr)
```

## Methods
There are three methods I want targeted

1. Iterating and returning the filename
2. Iterating and returning an ImageHDU
3. Iterating and returning an Array

These may seem simple right now, but I have lots of ideas for future improvements that will be based off this interface. In terms of implementation, I want to use `Task`s to create generators [this blog post](http://slendermeans.org/julia-iterators.html) shows a good example using a Task as a generator, and the Julia docs explain the functions as well. 

The reason I want to use Tasks is because we want to be able to modify certain data outside of the iteration- for instance, you'll want to do 
```julia
hdu = FITS(row.path)[row.hdu]
try # not sure if this is exact
    produce(hdu)
finally
	close(hdu)
end
```
to make sure the fits files close properly.

### PR DESCRIPTION
**Pull request by siddharthlal25**
This PR adds basic generators interface for iterating on file collection. Related issue: #25

### CODE REVIEW

**Comment by siddharthlal25** [comment]
How to document `@resumable` functions? When I added a block for documentation, it throws an error: 
```
cannot document the following expression:

#= /home/siddharth/.julia/dev/CCDReduction.jl/src/collection.jl:105 =# @resumable function array(df::DataFrame)
```

File: test/collection.jl [hunk 1]
```diff
@@ -45,6 +45,39 @@ using CCDReduction: parse_name
...
+@testset "image-generators" begin
+    # setting initial data
+    dir = joinpath(@__DIR__, "data")
+    df = fitscollection(dir)
+    arr1 = map(x->x, images(df))
+    arr2 = map(eachrow(df)) do row
+        FITS(row.path)[row.hdu]
+    end
+    @test arr1 == arr2
+end
```
Comment (siddharthlal25) [inline test/collection.jl]: I can't understand why this test fails, in my machine it shows the following stack:
```
image-generators: Test Failed at /home/siddharth/.julia/dev/CCDReduction.jl/test/collection.jl:78
  Expression: arr1 == arr2
   Evaluated: ImageHDU[File: /home/siddharth/.julia/dev/CCDReduction.jl/test/data/M35070V.fits
HDU: 1
Type: Image
Datatype: Int16
Datasize: (1059, 1059), File: /home/siddharth/.julia/dev/CCDReduction.jl/test/data/M6707HH.fits
HDU: 1
Type: Image
Datatype: Int16
Datasize: (1059, 1059)] == ImageHDU[File: /home/siddharth/.julia/dev/CCDReduction.jl/test/data/M35070V.fits
HDU: 1
Type: Image
Datatype: Int16
Datasize: (1059, 1059), File: /home/siddharth/.julia/dev/CCDReduction.jl/test/data/M6707HH.fits
HDU: 1
Type: Image
Datatype: Int16
Datasize: (1059, 1059)]
```
PS: I think the left side and right side are equal.

File: src/collection.jl [hunk 2]
```diff
@@ -95,3 +95,32 @@ function fitscollection(basedir::String;
...
+
+
+# generator for image arrays specified by data frames (i.e. path of file, hdu etc.)
+@resumable function array(df::DataFrame)
+    for row in eachrow(df)
+        fh = FITS(row.path)
+        try
+            @yield getdata(fh[row.hdu])
+        finally
+            close(fh)
```
Comment (giordano) [inline src/collection.jl]: Why not
```suggestion
        FITS(row.path) do fh
            @yield getdata(fh[row.hdu])
```
?

File: src/collection.jl [hunk 3]
```diff
@@ -95,3 +95,32 @@ function fitscollection(basedir::String;
...
+        @yield row.path
+    end
+end
+
+
+# generator for ImageHDU specified by data frame (i.e. path of file, hdu etc.)
+@resumable function images(df::DataFrame)
+	for row in eachrow(df)
+		@yield FITS(row.path)[row.hdu]
+	end
```
Comment (giordano) [inline src/collection.jl]: Indentation is different from the previous function

File: test/collection.jl (hunk 1 above)
Comment (giordano) [reply test/collection.jl]: `ImageHDU` is a mutable struct

File: src/collection.jl (hunk 2 above)
Comment (siddharthlal25) [reply src/collection.jl]: This one throws an error:
```
syntax: label "_STATE_1" referenced but not defined
```

Comment (giordano) [reply src/collection.jl]: Looks like a bug in `ResumableFunctions.jl`: https://github.com/BenLauwens/ResumableFunctions.jl/issues/28

File: src/CCDReduction.jl [hunk 4]
```diff
@@ -18,7 +19,10 @@ export subtract_bias,
        combine,
        subtract_dark,
        subtract_dark!,
-       fitscollection
+       fitscollection,
+       array,
+       filename,
+       images
```
Comment (mileslucas) [inline src/CCDReduction.jl]: we should use plural for all of these functions- `images` is an iterator of images, so we use plural. `array` is an iterator of arrays, yet it's being labeled `array`.

On top of that, `array` and `filename` as function names is just asking for confusion/name conflicts. `arrays` and `filenames` is a little less likely.

File: test/collection.jl [hunk 5]
```diff
@@ -45,6 +45,48 @@ using CCDReduction: parse_name
     @test size(df) == (0, 0)
 end
 
+@testset "array-generators" begin
+    # setting initial data
+    dir = joinpath(@__DIR__, "data")
+    df = fitscollection(dir)
+    arr1 = map(x -> x, array(df))
```
Comment (mileslucas) [inline test/collection.jl]:
```suggestion
    arr1 = array(df) |> collect
```

File: test/collection.jl [hunk 6]
```diff
@@ -45,6 +45,48 @@ using CCDReduction: parse_name
...
+        getdata(FITS(row.path)[row.hdu])
+    end
+    @test arr1 == arr2
+end
+
+@testset "filename-generators" begin
+    # setting initial data
+    dir = joinpath(@__DIR__, "data")
+    df = fitscollection(dir)
+    arr1 = map(x -> x, filename(df))
```
Comment (mileslucas) [inline test/collection.jl]: similar to above, try just using `collect`

**Review by mileslucas — CHANGES_REQUESTED** [review]
the core functionality looks good, needs docs and some changes first.

**Review by giordano — APPROVED** [review]
Looks good :+1:

File: src/collection.jl [hunk 7]
```diff
@@ -95,3 +95,53 @@ function fitscollection(basedir::String;
...
+
+# generator for image arrays specified by data frames (i.e. path of file, hdu etc.)
+@resumable function arrays(df::DataFrame)
+    for row in eachrow(df)
+        fh = FITS(row.path)
+        try
+            @yield getdata(fh[row.hdu])
+        finally
+            close(fh)
+        end
```
Comment (mileslucas) [inline src/collection.jl]: Does this work? The more file IO safeguards we can alleviate to other libraries, the better.
```suggestion
        FITS(row.path) do fh
            @yield getdata(fh[row.hdu])
        end
```

If this doesn't work, I'd vote to mimick the other two functions and skip the `try` block (just for consistency).

**Review by mileslucas — APPROVED** [review]
This looks good for the most part!

File: src/collection.jl (hunk 7 above)
Comment (siddharthlal25) [reply src/collection.jl]: This won't work, this is a bug from `ResumableFunctions.jl`.

**Pull request merged**